sphinx/util/inspect.py

```diff
@@ -249,62 +292,80 @@ def is_cython_function_or_method(obj: Any) -> bool:
 
 
 def isattributedescriptor(obj: Any) -> bool:
-    """Check if the object is an attribute like descriptor."""
+    """Check if the object is an attribute-like descriptor."""
     if inspect.isdatadescriptor(obj):
         # data descriptor is kind of attribute
         return True
     if isdescriptor(obj):
         # non data descriptor
         unwrapped = unwrap(obj)
-        if isfunction(unwrapped) or isbuiltin(unwrapped) or inspect.ismethod(unwrapped):
+        if isfunction(unwrapped) or isbuiltin(unwrapped) or ismethod(unwrapped):
             # attribute must not be either function, builtin and method
             return False
         if is_cython_function_or_method(unwrapped):
             # attribute must not be either function and method (for cython)
             return False
-        if inspect.isclass(unwrapped):
+        if isclass(unwrapped):
             # attribute must not be a class
             return False
-        if isinstance(unwrapped, (ClassMethodDescriptorType,
-                                  MethodDescriptorType,
-                                  WrapperDescriptorType)):
+        if isinstance(
+            unwrapped, (ClassMethodDescriptorType, MethodDescriptorType, WrapperDescriptorType)
+        ):
             # attribute must not be a method descriptor
             return False
         # attribute must not be an instancemethod (C-API)
-        return type(unwrapped).__name__ != "instancemethod"
+        return type(unwrapped).__name__ != 'instancemethod'
     return False
 
 
 def is_singledispatch_function(obj: Any) -> bool:
-    """Check if the object is singledispatch function."""
-    return (inspect.isfunction(obj) and
-            hasattr(obj, 'dispatch') and
-            hasattr(obj, 'register') and
-            obj.dispatch.__module__ == 'functools')
+    """Check if the object is a :func:`~functools.singledispatch` function."""
+    return (
+        inspect.isfunction(obj)
+        and hasattr(obj, 'dispatch')
+        and hasattr(obj, 'register')
+        and obj.dispatch.__module__ == 'functools'
+    )
 
 
-def is_singledispatch_method(obj: Any) -> bool:
-    """Check if the object is singledispatch method."""
+def is_singledispatch_method(obj: Any) -> TypeGuard[singledispatchmethod]:
+    """Check if the object is a :class:`~functools.singledispatchmethod`."""
     return isinstance(obj, singledispatchmethod)
 
 
-def isfunction(obj: Any) -> bool:
-    """Check if the object is function."""
+def isfunction(obj: Any) -> TypeGuard[types.FunctionType]:
```

Comment:
For this one, I'm not sure whether I should make it like that or simply type-guard it as a Callable (since we are unwrapping objects).